### PR TITLE
Closes #2964. z_sendmany once again makes v1 tx for taddr to taddr.

### LINKED ISSUE #2964: z_sendmany no longer creates v1 transactions for t->t, in violation of a SHOULD
@str4d wrote in a [comment](https://github.com/zcash/zcash/pull/2925#discussion_r167727060) on #2925:
> A side-effect of this is that ``z_sendmany`` no longer creates v1 transactions for t->t. This should be fine though, as anyone who has made the effort to use ``z_sendmany`` should already support parsing v2 transactions.

But this violates a SHOULD in section 6.1 of the protocol spec:
> Software that creates transactions **SHOULD** use version 1 for transactions with no JoinSplit descriptions.

This was motivated by maximizing compatibility with Bitcoin transaction parsing code. Note that @str4d's comment doesn't take into account that in general it will be different code (perhaps run by an entirely different party) parsing the transaction, than that calling ``z_sendmany``.

Either the implementation should be fixed or the spec should be relaxed.

### PR DESCRIPTION
**Pull request by bitcartel**

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3533,9 +3533,9 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
     UniValue contextInfo = o;
 
     // Contextual transaction we will build on
-    CMutableTransaction contextualTx = CreateNewContextualCMutableTransaction(
-        Params().GetConsensus(), chainActive.Height() + 1);
-    if (contextualTx.nVersion == 1) {
+    CMutableTransaction contextualTx = CreateNewContextualCMutableTransaction(Params().GetConsensus(), chainActive.Height() + 1);
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp]: Non-blocking: no need to change this line.

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Rebased on master. Here's the diff to the previously-reviewed commits, showing only the recent changes to master:
```diff
diff --git a/doc/wallet-backup.md b/doc/wallet-backup.md
index 9f7bc0f..02454db 100644
--- a/doc/wallet-backup.md
+++ b/doc/wallet-backup.md
@@ -18,7 +18,7 @@ There are multiple ways to make sure you have at least one other copy of the pri
 
 For all methods, you will need to include an export directory setting in your config file (`zcash.conf` located in the data directory which is `~/.zcash/` unless it's been overridden with `datadir=` setting):
 
-`exportdir=/path/to/chosen/export/directory`
+`exportdir=path/to/chosen/export/directory`
 
 You may chose any directory within the home directory as the location for export & backup files. If the directory doesn't exist, it will be created.
 
@@ -46,7 +46,7 @@ This will generate a file in the export directory listing all transparent and sh
 
 To import keys into a wallet which were previously exported to a file, use:
 
-`zcash-cli z_importwallet </path/to/exportdir/nameofbackup>`
+`zcash-cli z_importwallet <path/to/exportdir/nameofbackup>`
 
 ### Using `z_exportkey`, `z_importkey`, `dumpprivkey` & `importprivkey`
 
diff --git a/qa/rpc-tests/wallet_overwintertx.py b/qa/rpc-tests/wallet_overwintertx.py
index 6b65f21..176e869 100755
--- a/qa/rpc-tests/wallet_overwintertx.py
+++ b/qa/rpc-tests/wallet_overwintertx.py
@@ -43,6 +43,10 @@ class WalletOverwinterTxTest (BitcoinTestFramework):
         #
         # Currently at block 198. The next block to be mined 199 is a Sprout block
         #
+        bci = self.nodes[0].getblockchaininfo()
+        assert_equal(bci['consensus']['chaintip'], '00000000')
+        assert_equal(bci['consensus']['nextblock'], '00000000')
+        assert_equal(bci['upgrades']['5ba81b19']['status'], 'pending')
 
         # Node 0 sends transparent funds to Node 2
         tsendamount = Decimal('1.0')
@@ -85,6 +89,10 @@ class WalletOverwinterTxTest (BitcoinTestFramework):
         #
         # Currently at block 199. The next block to be mined 200 is an Overwinter block
         #
+        bci = self.nodes[0].getblockchaininfo()
+        assert_equal(bci['consensus']['chaintip'], '00000000')
+        assert_equal(bci['consensus']['nextblock'], '5ba81b19')
+        assert_equal(bci['upgrades']['5ba81b19']['status'], 'pending')
 
         # Node 0 sends transparent funds to Node 3
         tsendamount = Decimal('1.0')
@@ -104,9 +112,14 @@ class WalletOverwinterTxTest (BitcoinTestFramework):
         myopid = self.nodes[0].z_sendmany(taddr0, recipients)
         txid_shielded = wait_and_assert_operationid_status(self.nodes[0], myopid)
 
+        # Mine the first Overwinter block
         self.sync_all()
         self.nodes[0].generate(1)
         self.sync_all()
+        bci = self.nodes[0].getblockchaininfo()
+        assert_equal(bci['consensus']['chaintip'], '5ba81b19')
+        assert_equal(bci['consensus']['nextblock'], '5ba81b19')
+        assert_equal(bci['upgrades']['5ba81b19']['status'], 'active')
 
         # Verify balance
         assert_equal(self.nodes[1].z_getbalance(taddr1), Decimal('1.0'))
diff --git a/src/rpcblockchain.cpp b/src/rpcblockchain.cpp
index bbed8cb..626c472 100644
--- a/src/rpcblockchain.cpp
+++ b/src/rpcblockchain.cpp
@@ -659,12 +659,45 @@ static UniValue SoftForkDesc(const std::string &name, int version, CBlockIndex*
     return rv;
 }
 
+static UniValue NetworkUpgradeDesc(const Consensus::Params& consensusParams, Consensus::UpgradeIndex idx, int height)
+{
+    UniValue rv(UniValue::VOBJ);
+    auto upgrade = NetworkUpgradeInfo[idx];
+    rv.push_back(Pair("name", upgrade.strName));
+    rv.push_back(Pair("activationheight", consensusParams.vUpgrades[idx].nActivationHeight));
+    switch (NetworkUpgradeState(height, consensusParams, idx)) {
+        case UPGRADE_DISABLED: rv.push_back(Pair("status", "disabled")); break;
+        case UPGRADE_PENDING: rv.push_back(Pair("status", "pending")); break;
+        case UPGRADE_ACTIVE: rv.push_back(Pair("status", "active")); break;
+    }
+    rv.push_back(Pair("info", upgrade.strInfo));
+    return rv;
+}
+
+void NetworkUpgradeDescPushBack(
+    UniValue& networkUpgrades,
+    const Consensus::Params& consensusParams,
+    Consensus::UpgradeIndex idx,
+    int height)
+{
+    // Network upgrades with an activation height of NO_ACTIVATION_HEIGHT are
+    // hidden. This is used when network upgrade implementations are merged
+    // without specifying the activation height.
+    if (consensusParams.vUpgrades[idx].nActivationHeight != Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT) {
+        networkUpgrades.push_back(Pair(
+            HexInt(NetworkUpgradeInfo[idx].nBranchId),
+            NetworkUpgradeDesc(consensusParams, idx, height)));
+    }
+}
+
 UniValue getblockchaininfo(const UniValue& params, bool fHelp)
 {
     if (fHelp || params.size() != 0)
         throw runtime_error(
             "getblockchaininfo\n"
             "Returns an object containing various state info regarding block chain processing.\n"
+            "\nNote that when the chain tip is at the last block before a network upgrade activation,\n"
+            "consensus.chaintip != consensus.nextblock.\n"
             "\nResult:\n"
             "{\n"
             "  \"chain\": \"xxxx\",        (string) current network name as defined in BIP70 (main, test, regtest)\n"
@@ -687,7 +720,19 @@ UniValue getblockchaininfo(const UniValue& params, bool fHelp)
             "        },\n"
             "        \"reject\": { ... }      (object) progress toward rejecting pre-softfork blocks (same fields as \"enforce\")\n"
             "     }, ...\n"
-            "  ]\n"
+            "  ],\n"
+            "  \"upgrades\": {                (object) status of network upgrades\n"
+            "     \"xxxx\" : {                (string) branch ID of the upgrade\n"
+            "        \"name\": \"xxxx\",        (string) name of upgrade\n"
+            "        \"activationheight\": xxxxxx,  (numeric) block height of activation\n"
+            "        \"status\": \"xxxx\",      (string) status of upgrade\n"
+            "        \"info\": \"xxxx\",        (string) additional information about upgrade\n"
+            "     }, ...\n"
+            "  },\n"
+            "  \"consensus\": {               (object) branch IDs of the current and upcoming consensus rules\n"
+            "     \"chaintip\": \"xxxxxxxx\",   (string) branch ID used to validate the current chain tip\n"
+            "     \"nextblock\": \"xxxxxxxx\"   (string) branch ID that the next block will be validated under\n"
+            "  }\n"
             "}\n"
             "\nExamples:\n"
             + HelpExampleCli("getblockchaininfo", "")
@@ -722,6 +767,17 @@ UniValue getblockchaininfo(const UniValue& params, bool fHelp)
     softforks.push_back(SoftForkDesc("bip65", 4, tip, consensusParams));
     obj.push_back(Pair("softforks",             softforks));
 
+    UniValue upgrades(UniValue::VOBJ);
+    for (int i = Consensus::UPGRADE_OVERWINTER; i < Consensus::MAX_NETWORK_UPGRADES; i++) {
+        NetworkUpgradeDescPushBack(upgrades, consensusParams, Consensus::UpgradeIndex(i), tip->nHeight);
+    }
+    obj.push_back(Pair("upgrades", upgrades));
+
+    UniValue consensus(UniValue::VOBJ);
+    consensus.push_back(Pair("chaintip", HexInt(CurrentEpochBranchId(tip->nHeight, consensusParams))));
+    consensus.push_back(Pair("nextblock", HexInt(CurrentEpochBranchId(tip->nHeight + 1, consensusParams))));
+    obj.push_back(Pair("consensus", consensus));
+
     if (fPruneMode)
     {
         CBlockIndex *block = chainActive.Tip();
diff --git a/src/txdb.cpp b/src/txdb.cpp
index 94054dc..26f0814 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -314,6 +314,9 @@ bool CBlockTreeDB::LoadBlockIndexGuts()
                 pindexNew->nTx            = diskindex.nTx;
                 pindexNew->nSproutValue   = diskindex.nSproutValue;
 
+                auto header = pindexNew->GetBlockHeader();
+                if (!CheckEquihashSolution(&header, Params()))
+                    return error("LoadBlockIndex(): CheckEquihashSolution failed: %s", pindexNew->ToString());
                 if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, Params().GetConsensus()))
                     return error("LoadBlockIndex(): CheckProofOfWork failed: %s", pindexNew->ToString());
 
diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
index 1177cfc..890a91e 100644
--- a/src/wallet/rpcdump.cpp
+++ b/src/wallet/rpcdump.cpp
@@ -234,11 +234,11 @@ UniValue z_importwallet(const UniValue& params, bool fHelp)
             "1. \"filename\"    (string, required) The wallet file\n"
             "\nExamples:\n"
             "\nDump the wallet\n"
-            + HelpExampleCli("z_exportwallet", "\"test\"") +
+            + HelpExampleCli("z_exportwallet", "\"nameofbackup\"") +
             "\nImport the wallet\n"
-            + HelpExampleCli("z_importwallet", "\"test\"") +
+            + HelpExampleCli("z_importwallet", "\"path/to/exportdir/nameofbackup\"") +
             "\nImport using the json rpc call\n"
-            + HelpExampleRpc("z_importwallet", "\"test\"")
+            + HelpExampleRpc("z_importwallet", "\"path/to/exportdir/nameofbackup\"")
         );
 
 	return importwallet_impl(params, fHelp, true);
@@ -257,11 +257,11 @@ UniValue importwallet(const UniValue& params, bool fHelp)
             "1. \"filename\"    (string, required) The wallet file\n"
             "\nExamples:\n"
             "\nDump the wallet\n"
-            + HelpExampleCli("dumpwallet", "\"test\"") +
+            + HelpExampleCli("dumpwallet", "\"nameofbackup\"") +
             "\nImport the wallet\n"
-            + HelpExampleCli("importwallet", "\"test\"") +
+            + HelpExampleCli("importwallet", "\"path/to/exportdir/nameofbackup\"") +
             "\nImport using the json rpc call\n"
-            + HelpExampleRpc("importwallet", "\"test\"")
+            + HelpExampleRpc("importwallet", "\"path/to/exportdir/nameofbackup\"")
         );
 
 	return importwallet_impl(params, fHelp, false);
```

@zkbot r+

**Comment by bitcartel** [comment]
How did you generate the diff?

Also I see how you're using ```diff to get Github to display colours - nice!

**Comment by str4d** [comment]
I generated the diff with `git diff bitcartel/2906_v1_tx_no_assert..HEAD` before pushing, to show the changes relative to the pre-rebased branch.

**Pull request merged**